pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/InvertedIndexAndDictionaryBasedForwardIndexCreator.java

```diff
@@ -0,0 +1,609 @@
+      }
+
+      // Calculate value index for each docId in the forward index value buffer
+      // Re-use forward index length buffer to store the value index for each docId, where value index is the index in
+      // the forward index value buffer where we should put next dictId for the docId
+      int forwardValueIndex = 0;
+      for (int docId = 0; docId < _numDocs; docId++) {
+        int length = getInt(_forwardIndexLengthBuffer, docId);
+        putInt(_forwardIndexLengthBuffer, docId, forwardValueIndex);
+        forwardValueIndex += length;
+      }
+
+      // Construct the forward index values buffer from the inverted index using the length buffer for index tracking
+      for (int dictId = 0; dictId < _cardinality; dictId++) {
+        ImmutableRoaringBitmap docIdsBitmap = invertedIndexReader.getDocIds(dictId);
+        int finalDictId = dictId;
+        docIdsBitmap.stream().forEach(docId -> {
+          int index = getInt(_forwardIndexLengthBuffer, docId);
+          putInt(_forwardIndexValueBuffer, index, finalDictId);
+          putInt(_forwardIndexLengthBuffer, docId, index + 1);
+          if (shouldTrackVarLengthMetadata) {
+            trackMaxRowLengthInBytes(dictionary, maxRowLengthInBytes, docId, finalDictId);
+          }
+        });
+      }
+
+      IndexCreationContext.Forward context =
+          IndexCreationContext.builder().withIndexDir(_segmentMetadata.getIndexDir())
+              .withColumnMetadata(_columnMetadata).withforwardIndexDisabled(false).withDictionary(_dictionaryEnabled)
+              .withTotalNumberOfEntries(_nextValueId).withMaxNumberOfMultiValueElements(maxNumberOfMultiValues[0])
+              .withMaxRowLengthInBytes(maxRowLengthInBytes[0]).withLengthOfLongestEntry(lengthOfLongestEntry[0])
+              .build().forForwardIndex(_chunkCompressionType, _indexLoadingConfig.getColumnProperties());
+
+      writeToForwardIndex(dictionary, context);
+
+      // Setup and return the metadata properties to update
+      Map<String, String> metadataProperties = new HashMap<>();
+      metadataProperties.put(getKeyFor(_columnName, HAS_DICTIONARY), String.valueOf(_dictionaryEnabled));
+      metadataProperties.put(getKeyFor(_columnName, DICTIONARY_ELEMENT_SIZE),
+          String.valueOf(_dictionaryEnabled ? _columnMetadata.getColumnMaxLength() : 0));
+      metadataProperties.put(getKeyFor(_columnName, MAX_MULTI_VALUE_ELEMENTS),
+          String.valueOf(maxNumberOfMultiValues[0]));
+      metadataProperties.put(getKeyFor(_columnName, TOTAL_NUMBER_OF_ENTRIES),
+          String.valueOf(_nextValueId));
+      return metadataProperties;
+    }
+  }
+
+  private void trackLengthOfLongestEntry(Dictionary dictionary, int[] lengthOfLongestEntry, int dictId) {
+    switch (_columnMetadata.getFieldSpec().getDataType().getStoredType()) {
+      case STRING:
+        lengthOfLongestEntry[0] = Math.max(dictionary.getStringValue(dictId).length(),
```

Comment:
Don't we need to track the length as physical `number of bytes (UTF-8 encoded)` as opposed to `String.length()` which reports number of characters

Reply:
oops, good catch! i had a note in my TODO list for this and forgot to update it. I've fixed it.

Reply:
I tried printing the value and walking through the debugger too and found that both values come out to be the same in my tests:

Example log that I added:

```
        LOGGER.warn("*********** old len: {}, new len: {}", dictionary.getStringValue(dictId).length(), dictionary.getStringValue(dictId).getBytes(UTF_8).length);

```

test logs:

```
19:17:05.932 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 18, new len: 18
19:17:05.939 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 7, new len: 7
19:17:05.939 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 15, new len: 15
19:17:05.940 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 6, new len: 6
19:17:05.940 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 6, new len: 6
19:17:05.940 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 4, new len: 4
19:17:05.941 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 26, new len: 26
19:17:05.941 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 8, new len: 8
19:17:05.941 WARN [InvertedIndexAndDictionaryBasedForwardIndexCreator] [main] *********** old len: 19, new len: 19
```

So essentially they calculate the same thing, each UTF8 character is stored as a single byte. Does that make sense?

Reply:
Yes thanks for confirming and fixing.